File: macos/Sources/MacDroid/Packet.swift
import Foundation

struct Packet {
    let type: String
    let body: [String: Any]

    init(type: String, body: [String: Any] = [:]) {
        self.type = type
        self.body = body
    }

    /// Raw JSON bytes (no trailing newline) — used before framing/encryption.
    func jsonData() -> Data? {
        let dict: [String: Any] = [
            "id": Int(Date().timeIntervalSince1970 * 1000),
            "type": type,
            "body": body,
        ]
        return try? JSONSerialization.data(withJSONObject: dict)
    }

    func encode() -> Data? {
        guard var data = jsonData() else { return nil }
        data.append(0x0A)
        return data
    }

    static func decode(_ line: Data) -> Packet? {
        guard
            let obj = try? JSONSerialization.jsonObject(with: line) as? [String: Any],
            let type = obj["type"] as? String
        else { return nil }
        return Packet(type: type, body: obj["body"] as? [String: Any] ?? [:])
    }
}

File: android/app/src/main/java/com/macdroid/app/Packet.kt
package com.macdroid.app

import org.json.JSONObject

data class Packet(val type: String, val body: JSONObject = JSONObject()) {

    fun encode(): String = JSONObject().apply {
        put("id", System.currentTimeMillis())
        put("type", type)
        put("body", body)
    }.toString()

    companion object {
        fun decode(line: String): Packet? = try {
            val obj = JSONObject(line)
            Packet(obj.getString("type"), obj.optJSONObject("body") ?: JSONObject())
        } catch (_: Exception) {
            null
        }
    }
}
